scalding-commons/src/main/scala/com/twitter/scalding/commons/thrift/TProtocolOrderedSerialization.scala

```diff
@@ -0,0 +1,37 @@
+package com.twitter.scalding.commons.thrift
+
+import com.twitter.scalding.serialization.OrderedSerialization
+import cascading.tuple.hadoop.io.BufferedInputStream
+import cascading.tuple.StreamComparator
+import org.apache.thrift.protocol.TCompactProtocol
+import org.apache.thrift.protocol.TBinaryProtocol
+import org.apache.thrift.TBase
+import java.io.{ InputStream, OutputStream }
+import org.apache.thrift.transport.TIOStreamTransport
+
+import com.twitter.scalding.serialization.JavaStreamEnrichments._
+
+private[thrift] abstract class TProtocolOrderedSerialization[T] extends OrderedSerialization[T] {
+  val minFieldId: Short
+
+  @transient protected lazy val factory = new TBinaryProtocol.Factory //  new TCompactProtocol.Factory
+
+  def hash(t: T) = t.hashCode
+
+  def compareBinary(lhs: InputStream, rhs: InputStream): OrderedSerialization.Result = {
+    val leftSize = lhs.readSize
+    val rightSize = rhs.readSize
+
+    val seekingLeft = lhs.markOrBuffer(leftSize)
+    val seekingRight = rhs.markOrBuffer(rightSize)
```

Comment:
in practice the Hadoop InputStreams appear to always be backed with byte arrays, so they all support mark and reset, and they support fast skip.

The algorithm here is: if the InputStream supports mark/reset (which hadoop and cascading's do), we can use that at no cost. If they don't we read that number of bytes into a buffer and continue on from there. So, some kind of stream from a network that might not support seeking back would be completely read into our outside buffer, which is returned to us (though Java is so complected, my guess is that most NetworkStreams also have some mark buffer to support this).

At the end of this compare, we will have read the size and skipped that much data (by reset + skip or by reading it directly into a buffer).